File: src/P044/p044.go
package problem044

import (
	"strconv"

	"github.com/sunilbpandey/project-euler/src/utils/go/math"
)

func Solve() string {
	pentagonals := make(map[int]bool)
	answer := 0
	for p := range math.PentagonalNumbers() {
		for q := range pentagonals {
			if _, found := pentagonals[p-q]; !found {
				continue
			}

			if _, found := pentagonals[p-2*q]; !found {
				continue
			}
			answer = p - 2*q
			break
		}
		if answer > 0 {
			break
		}
		pentagonals[p] = true
	}
	return strconv.Itoa(answer)
}

File: src/P044/p044.py
from src.utils.py.series import pentagonal_numbers


def solve() -> int:
    pentagonals: set[int] = set()
    for curr in pentagonal_numbers():
        for prev in pentagonals:
            if curr - prev in pentagonals and curr - 2 * prev in pentagonals:
                return curr - 2 * prev
        pentagonals.add(curr)
    raise RuntimeError("No solution found")
